### PR TITLE
Add group context modal to file view

### LINKED ISSUE #483: request view - file page - "go to group" button
when viewing a file, it's difficult to know where to put a comment for that file

### PR DESCRIPTION
**Pull request by bloodearnest**
- **Refactor request group template context**
- **Add view context modal to contents view**
- **Add small functional test for context modal**
Button here: image

Modal: 
image

Fixes #483

### CODE REVIEW

File: airlock/views/request.py [hunk 1]
```diff
@@ -287,13 +285,23 @@ def group_presenter(release_request, relpath, request):
     visibilities = release_request.get_writable_comment_visibilities_for_user(
         request.user
     )
+    # are we on the group page?
+    if len(relpath.parts) == 1:
+        inline = False
+    else:
+        inline = True
```
Comment (madwort) [inline airlock/views/request.py:292]: it's semi-tempting to do a refactor to add some convenience methods to `UrlPath` for things like this, although from a quick skim of the codebase we've not hit the 3-4 repeated usages I'd want to see to make me think it's worthwhile. (I'm thinking `relpath.is_group()` for here and also an accessor-type method `relpath.group` in place if `relpath.parts[0]`.

File: airlock/views/request.py [hunk 2]
```diff
@@ -323,12 +270,68 @@ def request_view(request, request_id: str, path: str = ""):
...
+
+    # are comments readonly?
+    c2_readonly = inline or not permissions.user_can_edit_request(
+        request.user, release_request
+    )
```
Comment (madwort) [inline airlock/views/request.py:297]: looks like c2 is context & controls, not comments?

```suggestion
    # are context and controls readonly?
    c2_readonly = inline or not permissions.user_can_edit_request(
        request.user, release_request
    )
```

Comment (madwort) [reply airlock/views/request.py:297]: (this is why I like extra-long variable names...)

File: airlock/templates/file_browser/group.html [hunk 3]
```diff
@@ -7,7 +7,7 @@
     background-color: darkgrey;
   }
 </style>
-{% #card title=path_item.name container=True %}
+{% #card title=group.title|default_if_none:path_item.name container=True class="group_modal" %}
```
Comment (madwort) [inline airlock/templates/file_browser/group.html:10]: in what context could `group.title` be none?

**Review by madwort — APPROVED** [review]

File: airlock/views/request.py (hunk 1 above)
Comment (bloodearnest) [reply airlock/views/request.py:292]: Well, that's a whole thing

There are times when a UrlPath means "path to file in a workspace" and times when it means "actual url" and times where it means "request url that will include group". cf `bll.get_request_file_from_output_path and bll.get_request_file_from_urlpath`.

Previously, I tried to split those into separate types: OutputPath, UrlPath, RequestUrlPath (which had a `.group` as you suggest and a `.output_path` w/o the group), and use the right types to help us catch these distintions.

But at the time, I couldn't make it work cleanly.  It's probably worth trying again, and definitely would be a good refactor.

File: airlock/templates/file_browser/group.html (hunk 3 above)
Comment (bloodearnest) [reply airlock/templates/file_browser/group.html:10]: ah, it was possibly none at an earlier point in the branch, when I added this. Will remove.

Comment (bloodearnest) [reply airlock/templates/file_browser/group.html]: fixed

File: airlock/views/request.py (hunk 2 above)
Comment (madwort) [reply airlock/views/request.py]: > Curious - you would seriously prefer context_and_controls_readonly here?

yes I would! I know I'm an outlier...

File: airlock/views/request.py (hunk 1 above)
Comment (madwort) [reply airlock/views/request.py:292]: yeah, I think it would be a really valuable refactor, if we have time.

Comment (madwort) [reply airlock/views/request.py:292]: put a ticket on the triage pile https://github.com/opensafely-core/airlock/issues/622

**Pull request merged**